src/main/java/reactor/core/publisher/FluxBufferPredicate.java

```diff
@@ -0,0 +1,295 @@
+ *     non-matching element is simply discarded.</li>
+ * </ul>
+ *
+ * @param <T> the source value type
+ * @param <C> the output collection type
+ * @see <a href="https://github.com/reactor/reactive-streams-commons">Reactive-Streams-Commons</a>
+ */
+final class FluxBufferPredicate<T, C extends Collection<? super T>>
+		extends FluxSource<T, C> {
+
+	public enum Mode {
+		UNTIL, UNTIL_OTHER, WHILE;
+	}
+
+	final Predicate<? super T> predicate;
+
+	final Supplier<C> bufferSupplier;
+
+	final Mode mode;
+
+	public FluxBufferPredicate(Publisher<? extends T> source, Predicate<? super T> predicate,
+			Supplier<C> bufferSupplier, Mode mode) {
+		super(source);
+		this.predicate = Objects.requireNonNull(predicate, "predicate");
+		this.bufferSupplier = Objects.requireNonNull(bufferSupplier, "bufferSupplier");
+		this.mode = mode;
+	}
+
+	@Override
+	public long getPrefetch() {
+		return Long.MAX_VALUE;
+	}
+	
+	@Override
+	public void subscribe(Subscriber<? super C> s) {
+		C initialBuffer;
+		
+		try {
+			initialBuffer = bufferSupplier.get();
+		} catch (Throwable e) {
+			Operators.error(s, Operators.onOperatorError(e));
+			return;
+		}
+		
+		if (initialBuffer == null) {
+			Operators.error(s, new NullPointerException("The bufferSupplier returned a null initial buffer"));
+			return;
+		}
+
+		BufferPredicateSubscriber<T, C> parent = new BufferPredicateSubscriber<>(s,
+				initialBuffer, bufferSupplier, predicate, mode);
+
+		s.onSubscribe(parent);//?
+		source.subscribe(parent);
+	}
+	
+	static final class BufferPredicateSubscriber<T, C extends Collection<? super T>>
+	implements Subscriber<T>, Subscription {
+
+		final Subscriber<? super C> actual;
+		
+		final Supplier<C> bufferSupplier;
+
+		final Predicate<? super T> predicate;
+
+		final Mode mode;
+
+		C buffer;
+		
+		volatile Subscription s;
+		@SuppressWarnings("rawtypes")
+		static final AtomicReferenceFieldUpdater<BufferPredicateSubscriber, Subscription> S =
+				AtomicReferenceFieldUpdater.newUpdater(BufferPredicateSubscriber.class, Subscription.class, "s");
+		
+		volatile long requested;
+		@SuppressWarnings("rawtypes")
+		static final AtomicLongFieldUpdater<BufferPredicateSubscriber> REQUESTED =
+				AtomicLongFieldUpdater.newUpdater(BufferPredicateSubscriber.class, "requested");
+		
+		BufferPredicateSubscriber(Subscriber<? super C> actual, C initialBuffer,
+				Supplier<C> bufferSupplier, Predicate<? super T> predicate, Mode mode) {
+			this.actual = actual;
+			this.buffer = initialBuffer;
+			this.bufferSupplier = bufferSupplier;
+			this.predicate = predicate;
+			this.mode = mode;
+		}
+		
+		@Override
+		public void request(long n) {
+			if (Operators.validate(n)) {
+				Operators.getAndAddCap(REQUESTED, this, n);
+			}
+		}
+
+		@Override
+		public void cancel() {
+			Operators.terminate(S, this);
+		}
+
+		@Override
+		public void onSubscribe(Subscription s) {
+			if (Operators.setOnce(S, this, s)) {
+				s.request(Long.MAX_VALUE);
```

Comment:
I had that in mind, yes, to request no more than the necessary number of elements needed to trigger a buffer boundary... I went with the minimum viable product for now but I'll see if I can quickly amend the change.